Network Buffer FFM & Network Conformance

kermandev opened the pull request:
## Proposed changes
Tracking PR for FFM, being a squash of all existing FFM work; See #2677 for more history on changes.
Description will be updated with changes and reasoning.

# FFM Changelog

## NetworkBuffer
* `NetworkBuffer` is unsealed and documented.
* **Type Updates**:
  * `OPTIONAL_NBT` and `OPTIONAL_NBT_COMPOUND` types to cleanly handle optional NBT data without arbitrary branching.
  * `LP_ANGLE` maps 0-360 degrees to a single unsigned byte.
  * `FixedBitSet` and `EnumSet` write proper fixed bitsets.
  * `Type#unionType` uses a single generic bound.
  * `BLOCK_POSITION` now returns `BlockVec` instead of `Vec`.
  * `Recursive`: Added `Recursive` similar to `Codec#Recursive`
* **Management**:
  * `#requestCapacity(long)` added for auto resizing.
  * `#ensureReadable(long)` fails if there aren't enough bytes to read.
  * `#trim()` and `#trimmed(...)` added to compact and shrink resizable buffers.
  * `#slice(...)` returns a new instance backed by a slice of the parent buffer.
  * `#copy(NetworkBufferAllocator, ...)` overloads copy for a specific allocator.
  * `#readOnly()` returns a read only *view*. If the segment is resizeable, when any operation is attempted, the read only view is not required to be the same.
  * `#fill(...)` added along with zeroing documentation.
* **Deprecations & Removals**:
  * `#extractBytes()` deprecated for `#extractReadBytes(...)` or `#extractWrittenBytes(...)`.
  * `#copyTo(long, byte[], long, long)` deprecated.
  * `#equals` deprecated for `#contentEquals`.
* **Usage**:
  * `staticBuffer(..., Registries)` and `resizableBuffer(..., Registries)` now require `Registries` to be non null.
  * `makeArray(Type, T)` has been added as a shortcut. Existing `makeArray(..., Registries)` still accepts a nullable `Registries` to easily allow passing through registries from another `NetworkBuffer` without manual null checks. Note that `makeArray` now uses a confined arena, meaning the returned buffer is strictly bound to the thread that initially created it and will throw an exception if accessed by other threads.
  * `wrap(MemorySegment, ...)` was added as a new experimental feature.
  * `dummy(Registries)` was introduced to return a dummy buffer (useful for `Type#sizeOf`) and uses a nullable `Registries`.

## NetworkBuffer Additions
* **Advanced Memory Allocation API**:
  * `NetworkBufferAllocator`: New functional interface defining how `NetworkBuffer` objects should be allocated for network buffers. Supports explicit `Arena` assignments (`staticAllocator()` or `resizeableAllocator()`).
  * `NetworkBufferSegmentProvider`: Not really intended for general API use yet, but allows advanced users a way to access the backing `MemorySegment` of a buffer.
  * **Native Memory**: Uses native memory allocation (`malloc` and `free`) bypassing standard Java memory clearing when `ServerFlag.ATTEMPT_NATIVE_ALLOCATION` is true.
* **Views**:
  * `NetworkBuffer.IOView`: Extends `DataInput` and `DataOutput` to mostly interface with Java's streams, used with Adventure.
  * `NetworkBuffer.Direct`: Used internally denoting that interactions write directly to the buffer, ignoring indiences rules.

## Reading & Writing
* **Slices**: Packet reading uses read only slices to stop readers from messing with the buffer.
* **Error Handling**: Restructures how invalid packets are handled and logged using server flags:
  * **Packet Misuse**: If a packet handler throws an exception during execution, it triggers a disconnect (`REJECT_MISUSED_PACKET`).
  * **Packet Malformed**: If a packet constructor or any point during reading throws an exception (invalid data), it triggers a disconnect (`REJECT_MALFORMED_PACKET`).
  * **Error Suppression**: Added `SUPPRESS_PACKET_ERROR_LEVEL` to hide packet reading exceptions based on the `ConnectionState` ordinal. This is extremely useful for preventing console log spam from automated server scanners and pingers. Handled securely inside `PlayerSocketConnection` and `ExceptionManager`.
* **Connection Resets**: The server writer ignores "Connection Reset by peer" errors.
* **Disconnection Leak**: Fixed the leak of `PlayerSocketConnection`s when a connection is closed using the new write lock.

## Memory & Caching
* **BufferedPacket**: Sets it's buffer to a read only view in the canonical constructor.
* **CachedPacket** & **LazyPacket**: Uses weaker acquire/release semantics to be better on ARM platforms.
* **LazyPacket**: Now uses synchronized in the slow path, compared to racing till the last winner, as virtual thread pinning has been fixed in JDK 25.
* **FramedPacket**: Its `readIndex` gets set after going read only in the constructor.
* **SendablePacket**: `#extractServerPacket` requires a writer for cached packets now, and may require different fields in the future.

## Packets
* **Action based**: `StopSoundPacket` and `ScoreboardObjectivePacket` are now action based (e.g., `new StopSoundPacket(new StopSoundPacket.All())` instead of `new StopSoundPacket(0, null, null)`).
* **PlayerInfoUpdatePacket**: `PlayerInfoUpdatePacket.Property` removed in favor of `GameProfile.Property`.
* **DamageEventPacket**: Updated to use the registry with nullable attacker and source entity IDs.
* **FacePlayerPacket**: Entity data now uses a nullable `EntityData` instead of special values (e.g `new FacePlayerPacket(FacePosition.EYES, Vec.ZERO, null)` instead of `new FacePlayerPacket(FacePosition.EYES, Vec.ZERO, 0, null)`).
* **StatusRequestPacket**: Renamed to `ClientStatusRequestPacket`.
* **LegacyServerListPingPacket**: `LegacyServerListPingPacket`, `ServerListPingType`, and related methods are gone. These packets have not been handled for quite a while now.
* **ChunkData**: Chunk data is now formed using `BlockEntityInfo`.
* **MagicConstants**: Added flag support to `ClientPlayerPositionAndRotationPacket`, `ClientPlayerPositionPacket`, `ClientPlayerPositionStatusPacket`, `ClientPlayerRotationPacket`, `ClientUpdateStructureBlockPacket`, `RespawnPacket`, and `DeclareCommandsPacket.Node`.
* **ComponentHolding**: Added to `ServerLinksPacket` (which now uses `Either` for Entry), `MapDataPacket`, `PlaceGhostRecipePacket`, and `SetTitleTextPacket` / `SetTitleSubTitlePacket`.
* **ParticlePacket**: Origins and offsets now use `Point`. The old constructors are still available for compatibility but are now deprecated.
* **SoundEffectPacket**: Translates values locally wrapping divisions implicitly inside protocol calls.
* **Templates**: Packets that used to have a manual type are now using templates.
* **Immutability**: Packets should no longer be compared by identity and are generally more immutable overall (e.g., clone applied on byte arrays in `ClientPluginMessagePacket`, `ClientEncryptionResponsePacket`, `ClientLoginPluginResponsePacket`, `MultiBlockChangePacket`, `RegistryDataPacket`, etc).

## New Server Flags
* `SUPPRESS_CONNECTION_ACCEPT_ERRORS`: Suppresses traces caught explicitly during the server's main socket accept phase when a client first connects (default: `true`). Useful to suppress scanners.
* `REJECT_MALFORMED_PACKET`: Drops connections if a packet constructor or reading parsing throws an exception (default: `true`).
* `REJECT_MISUSED_PACKET`: Drops connections if a packet handler throws an exception during execution (default: `true`). This may be unwanted from current behavior if you want players to still be connected if their packet was mishandled.
* `SUPPRESS_PACKET_ERROR_LEVEL`: Suppresses packet reading errors thresholded by >= `ConnectionState` (default: `-1`, no supression).
* `WARN_UNREAD_BYTES_PACKET`: Triggers a console warning if a packet is read but unread bytes remain in the payload (default: `true`).
* `AUTOMATIC_COMPONENT_TRANSLATION`: Automatically attempts to translate components server side before sending (default: `false`). This replaced `MinstomAdventure.AUTOMATIC_COMPONENT_TRANSLATION`.
* `DEFAULT_RESIZEABLE_SIZE`: The baseline size of newly allocated resizable NetworkBuffers (default: `256`).
* `ATTEMPT_NATIVE_ALLOCATION` / `FORCE_NATIVE_ALLOCATION`: Used to bypass standard Arena `calloc` zeroing by directly calling `malloc` and `free` using native memory segments for better performance (defaults: `true` / `false`). Currently only supports 64 bit machines. 
* `DECOMPRESS_POOL_SIZE` / `COMPRESS_POOL_SIZE` / `PACKET_POOL_SIZE` / `VIEWABLE_POOL_SIZE`: Allows configuring the size of various internal memory pools that were previously unconfigurable (default: `32767`).

## Miscellaneous
* **Damage API**: `Damage#getTypeId()` is deprecated in favor of `#getType()`.
* **Codec BlockVec**: `BLOCK_POSITION` Codec now returns `BlockVec` to align with NetworkBuffer's.
* **ComponentSerialization**: Transitioned from `NetworkBufferTypeImpl` interface to returning instances of `NetworkBuffer.Type<Component>`. Lookups for registry payloads use the buffer's `ioView()`.
* **VarInt & VarLong**: Now can be used when the buffer is under 5/10 bytes.
* **VarInt**: Packets using small `VarInt`s (0-127) now serialize using `Byte` instead. This is purely unnoticeable internally and no types should change.
* **MessageSignature**: Fixed to strictly 256 bytes (no longer uses `RAW_BYTES`). `MessageSignature.Packed` network type is now properly implemented.
* **ObjectPool**: Split implementation (still internal), supports `Unpooled`, and moved to the `collections` package.
* **Binary Tags**: `BinaryTagReader` & `BinaryTagWriter` are now records around `DataInput`/`DataOutput`.
* **Packet Listeners**: Registrations in `PacketListenerManager` renamed from `StatusRequestPacket` to `ClientStatusRequestPacket` and updated consumers (`PacketPlayListenerConsumer`, `PacketPrePlayListenerConsumer`) to throw Exception out of `.accept()`.
* **JMH Benchmarks**: Added multiple network buffer, memory allocation, and packet benchmarks.
* **Module Exports**: `module-info` exports `net.minestom.server.network.foreign`.

## Types of changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Publishing
These changes are published so you can use them by accessing the snapshots repository.
```
repositories {
    maven(url = "https://central.sonatype.com/repository/maven-snapshots/") {
        content {
            includeModule("net.minestom", "minestom")
        }
    }
    mavenCentral()
}

dependencies {
    implementation("net.minestom:minestom:ffm-SNAPSHOT")
}
```

kermandev committed:
squash: Squash all changes

chore: Post rebase changes

chore: Update reachability metadata

refactor: Change to use PacketParser.Server for ServerPacket

revert: DataComponent changes

freeze is integrated on main, and poorly rebased.

fix: Post rebase changes

Alot of things to change, alot of diffs to go over after the force push is completed.

refactor: Use constants in VarInt/Varlong

Fix from the regression caused by 31d9e5d4a7576da33998e37b0ff184e6328880f1

Delete some sizeOf's that are complicated and dont capture the full behvaior like FixedByteArray or otptional + either logic can just be simplified into a single dummy. We shouldnt use sizeof for branching logic

refactor: VarInt & VarLong write logic.

Simpler and faster it seems. custom sizeOf slower than just writing it twice.

refactor: Drop F1 & F2 & SFN

SFN was used by templating, it has already been reverted.

fix: Assertions & tests

Assertions were backwords from what Check uses.

Also failed to migrate the VarLong right, and the sizeOf being recursive.

fix!: Some missing parts of reverts

Also fixes NetworkBuffer#resizableBuffer(int) to be a long.
Changes lengths to assert cause its useless at runtime for 99.999% of cases most likely.

revert: "fix: GraalVM native image"

This reverts commit 727af12e

Part of the goal is to trim protypes into other PRs as they can be reviewed easier, this is the FFM branch lets reduce the scope.

revert: "fix: Freeze collection based data component values"

This reverts commit 80ac7d70bde8f3ed4ab1f0e75d74fc3a2967556c.

Revert "fix: javadoc"

This reverts commit ff4e5c189eaa1eb5b1813aebf1b75f35b465eb3e.

Part of the goal is to trim protypes into other PRs as they can be reviewed easier, this is the FFM branch lets reduce the scope.

revert: Network template system

Part of the goal is to trim protypes into other PRs as they can be reviewed easier, this is the FFM branch lets reduce the scope.

fix: Disconnect on more exceptions

Fixes an issue where client exceptions counted as server exceptions

refactor: Change templates to get their own callers.

Changes templateN to do the null checks now, and callers must be passed through from template.

Also fixes some minor type issues.

fix: Suppress more templating from Native image reachability-metadata.json

Native images will require the user to build at runtime anyway, the metadata added for reflection will only increase the size of the final executable as its added metadata.

fix: Exclude generated classes from the trace

fix: GraalVM native image

Viewable still is a problem, but can be alleviated if you turn the pooling for Network buffers for viewable to zero.

Tested with:
native-image --initialize-at-build-time -march=native --enable-https -jar minestom-demo.jar -Dminestom.viewable-packet=false

fix!: Dont generate debug templates

chore!: Update metadata with new templating

This likely needs more looks as templating does a lot of runtime reflection to set up the classes, so we might need a fallback mode for graalvm.

feat!: Introduce network template system for runtime network type generation

Allows some very nice implementations that are even faster than some of the handwritten ones as we can directly use the NetworkBuffer.Direct api. Uses Hidden classes & ClassFile api, likely to be much faster than the existing users that omit the `-XX:+UnlockExperimentalVMOptions -XX:+TrustFinalNonStaticFields` from their startup flags, as Minestom gets much faster (as in peak performance) if you can use it.

Some features include: auto unboxing, and unboxing through transforms.
For example, a template that includes an Enum(A.class) is just emits a write instruction that uses an unboxed byte, well if A.values().length < 128, of course, VarInt's are also supported.
Some features that could be added are index offsets for constant size templates, (Ones without VarInt, and are just primitives to start with), and Optional inlining for primitives.

Some quite big hacks that cause breakage include using Serialization (SF1.SF2...SF19.SF20) to access the method handle that the lambda is consuming so we can avoid more boxing, quite a hack that will likely disappear with Valhalla.

Breaking: Any project using templates will require a recomp as the ABI is broken and will not be fixed. Same with #transform, but the legacy method can still exist, you just won't get any benefits using it.

There might be more breakages in the future depending on what the templating should inline. This also begs the question of templating singles like NetworkBuffer.Enum

refactor: Generics in PacketRegistry

Also adds ServerPacket.Handshake in case its used in the future.

refactor: PacketReader should pass through PacketRegistry.

fix: Possible failing pool tests

fix: Methods inside NetworkBufferDummy are too strict

refactor: Allow Server to change the packet writer.

fix: Pull outgoing handle into static final

fix: Module builds

refactor: Pass through sizeOf

Simple things that have a parent should probably pass through the sizeOf like transform, or Either

Should help avoid bad scalar replacement, for mostly constant things.

refactor: Move Dummy buffer into its own implementation.

During this change a lot of similar methods could just be moved into NetworkBuffer

The way NetworkBufferDummy is implemented does allow for it methods to be fully resolved after the potential vtable call of sizeOf, it has the added benefit of removing all the isDummy checks from NetworkBufferSegmentImpl, which does make the implementation nicer.

sizeOf checks should be able to be compiled into direct additions now.

fix: Demo server translation

fix: Missing block entity check

refactor: Unbound extendable classes

Makes bounds quite a generics quite a bit easier to use.

refactor: Make PlayerSocketConnection final

refactor: Delete NetworkBufferProvider.

The implementation seems half-baked, and with a server builder as alot of state could likely be removed in the future. We can revisit this later, but currently the auth flow needs to be reworked before a difentive implementation on how to allow custom network buffers is allowed.

refactor: Immutable DeclareCommandsPacket.Node

Why did Mojang decide to make this packet so unstructured makes me sad.

Also the GraphConverter code is very complex, sadly.

fix: unbound sendPackets & use compareAndSet

fix: Force an write thread unlock.

Haven't seen the case, but id rather it be spurious than not be.

chore: Update reachability-metadata.json

refactor: Some final packet changes

Breaking DamageEventPacket, no good way around it without doing a registry lookup.

Docs are still required, and will be a much bigger changelist

Refactor: Improve packet error/warn handling

Also fixes small edge cases & cleanup a bit

Refactor: move AUTOMATIC_COMPONENT_TRANSLATION to ServerFlag

refactor: Move ObjectPool into collection package

chore: optimize imports for network

refactor: Use Byte for small enums

Likely covers all cases

refactor: Allow nullable returns from templates

Should prevent @NotNull requirements for templates!

refactor: NetworkBufferSegmentAllocator.allocate

Also removes the checks performed in NetworkBufferResizeableSegment as they were wrong (NULL for zero) and isReadOnly doesnt need to be checked here.

refactor: Delete NetworkBufferTypeImpl.UnitType and use NetworkBufferTemplate

Also adds 0 sizeOf override.

refactor: Change how sizeOf primitives are calculated.

Fix test issues, as it's logically wrong.
Writing VarInts zenv4 are 0.4ns faster on my machine (needs more testing)
Reading performance is the same.

refactor: Use new singleton instances where applicable

Ignores tests, as these singletons will be removed once value types exist and new T() == new T()

refactor: Expose singleton instance for network buffer templates.

refactor: Relax NetworkBufferTemplate requirements

refactor: Revert making all registry calls not null.

Use JDK 27 string api (partial) when supported. More to come

They will still exist when creating a buffer, as intent is very important there.

Loosen bounds of network types, Very important to do more.

fix: Delete unsafe again

Dont need the implementation, used for testing.

fix: Dont use parallelStream on empty map

refactor: Remove more legacy ping support

Start comparing Status by its value instead of identity on array

fix: SegmentProvider javadoc

feat: refactor NetworkBuffer.Type implementations serialization and deserialization

Also changes how providers work to be a ServerFlag for now instead of a half baked init.

refactor: separate CachedPacket slow path

feat: dont force ordering when caching is disabled, finalize packet registry methods

fix: Broken network test

feat: replace Vec with BlockVec for reading block positions

feat: add support for custom network buffer provider in server init

feat: readd unsafe network buffer implementation

refactor: improve documentation, and remove invalid assert

Dont use volatile for packet caching in CachedPacket and LazyPacket

Fix version merge.

Required to add equals/hashcode to WeightedList to for tests

Update docs

Update test to not use deprecated constructor

Cleanup changes

Expose segment if applicable. Make factories no longer stored, and let the implementation perform it. We should always use the default implementation if required.

Post merge fixes

Add some more documentation on networkBufferProvider.

Also rename from #get to #networkBufferProvider

Perform some cleanup.

Abstract NetworkBuffer implementation into foreign

Should allow for a netty implementation if desired by users as packet writing/reading was the major blocker. The login process will not be touched here.

Rename to SegmentImpl

Fix tests

Wip

Remove hashcode/equals implementation for NetworkBuffer

These should be dependent on identity, due to how NetworkBuffer really cant be immutable due to tracking index sadly.

Also make requireCapacity checked with IndexOutOfBoundsException

Rename LP_ROTATION to LP_ANGLE

Makes it more generalized.

Kick player on any malformed reading

Also improves error messages

Fix string

Make NetworkBufferIOView check for nullability

Add additional biome packet test

Drop ByteEnum

Values are the same up to 127 entries, no reason to have a specialized type here.

Remove reachability fence

Not required here.

Add ObjectPool, 0 size support.

This should allow for better folding if there are no pools, and is somewhat simplier but kind of goes against the ObjectPool introducing a unpooled variant.

Flip final static to static final

MagicConstants for position/rotation packets

Structure Server packets more

Also fix some missing component holders and refactor ScoreboardObjectivePacket to an actual proper union instead of the weird special flag

Trying to convert everything into templates if applicable as its a better contract that a new Type

Add LP_FLOAT type

Structure Client packets more

Fix ClientPluginMessagePacket not copying, but it's unlikely it's ever necessary

Also convert InteractEntity to union and declare id internal.

Move PlayerHand enum into its class, as its used in both client and server and is duplicated a few places

Minor cleanup

Make the NetworkBufferAllocator much more robust

Put the size and address if available

Use a global cleaner

Rename cleaner to free

Add a reachability fence for the arena

Throw OutOfMemory outside the segment.

Also clean if it fails to reinterpret

Revert "Force encryption to be set on the read thread"

This reverts commit f0f406a3a1c1a1cece5946c71d872b7138df1dab.

Force encryption to be set on the read thread

Fix NetworkBufferAllocator zero case for NULL check

Man page for malloc states; `The behavior of these functions when the requested size is zero is glibc specific; other implementations may return NULL without setting errno`

Fix MessageSignature

Fix getBossBarFlagValue

Fix code block

Refactor NetworkBuffer documentation and remove unused imports

Revert copyTo layout

The layout may not be correct to encoding format of the same type; it exposes the memory segment to literary, without a custom factory.

Remove var from makeArray

Also revert sizeOf check for size

Update doc

Change Settings to Factory.

A better name for it, Settings assume they could be modified for a segment.

Inline arena creation

From old implementation

Change NetworkBuffer javadocs

Expose copyTo for more than just bytes.

Fix javadoc from fill change

Network Buffer FFM & more conformance (#2677)

A squash of all existing work; See #2677 for more history on changes.

mudkipdev commented on src/main/java/net/minestom/server/adventure/AdventurePacketConvertor.java:
Boss-bar flags are accumulated with `val |= flag.ordinal()`. The first flag (ordinal 0) never sets any bits, and the others set the wrong bit positions. Every boss-bar flag payload sent from these helpers is therefore wrong once multiple flags are involved.

kermandev committed:
feat!: Use JPMS instead of shadow jars.

This updates to 26.1-shapshot-11-dev (jpms) for data to get resource() API and drops shadowJar in favor of setting the module path under application.

kermandev committed:
feat: Add the new registries to data generator

kermandev committed:
chore: Run Datagenerator for 26_1 snapshot 11

kermandev committed:
feat: Add new registries to Minestom

kermandev committed:
feat!: Fix existing registries

kermandev committed:
feat!: Fix existing components & metadata

kermandev committed:
feat!: Fix & add new packets

kermandev committed:
feat: New SlotDisplay types

kermandev committed:
fix: (backport) Fix JPMS issues with net.kyori.examination.string

kermandev committed:
feat: Timeline ClockTimeMarker & WorldClock

This also moves WorldClock into clock folder.

kermandev committed:
chore: Update code generation to 26.1-pre-1

kermandev committed:
feat: Update to 26.1-pre-1

kermandev committed:
chore: Stop using Check.notNull(Object, String)

Instead, use Objects.requireNonNull

kermandev committed:
chore: Discourage Check.fail(String)

kermandev committed:
fix: IntProvider codec is recursive

Also changes CodecImpl.ForwardRefImpl#delegate to do a single field read

kermandev committed:
fix: Packet change for TimeUpdate

kermandev committed:
fix: Use the default clock for the dimension

kermandev committed:
fix: Add possible new tags to tags packet

kermandev committed:
chore: 26.1-pre-2

mworzala committed:
feat: update components to 26.1, bump data to rc-2

mworzala committed:
feat: clock tracking in instance

mworzala committed:
feat: clock tests

mworzala committed:
fix: DamageResistant is still wrapped in an object

mworzala committed:
fix: ItemStackTemplate encodes item -> slot, unlike ItemStack which is slot first

kermandev committed:
feat: Add GameRule registry & events

kermandev committed:
refactor!: Handle the new spectate packets

It's easier to classify these as new as the old functionality would be VERY broken, as teleport and spectate would be swapped.

kermandev committed:
fix(jpms): Fix testing jpms support

Testing needed to open its module for junit to allow it to use the environment.

Also fixes the failed export from worldevent

kermandev committed:
chore: Update dependencies

kermandev committed:
refactor: Use component instead for ZombieNautilusMeta

kermandev committed:
chore: Update test worlds

Uses 1.21.11 as the updater, as no new blocks need translation currently.

kermandev committed:
feat(testing): Ensure tests fail on ExceptionHandler by default.

kermandev committed:
feat: Update to the 26.1 release

kermandev committed:
chore: Remove some todos

SorkoPiko committed:
feat: `BlockInstanceEvent` (#3003)

* feat: BlockInstanceEvent

* refactor: make BlockEvent extend InstanceEvent

* refactor: use PlayerInstanceEvent over PlayerEvent

kermandev committed:
feat!: Packet State Markers (#3041)

* feat: Introduce ClientPacket.ConnectionState's

Adds markers for serialization buckets similar to ServerPacket.

* feat: Introduce ServerPacket.Handshake

Just to match all the connection states, even if its unused its missing.

* refactor!: Enforce PacketRegistry types for Client and Server.

Exposes Client & Server Directly as they are the only implementors and no reason to use the generic version.

Might be breaking, not for ABI likely but for anyone using the PacketParser directly to add packets, Unsure how widely used this is.

* fix: Dont introduce var into PacketWriting

* refactor!: Do the full abi break

bed-dev committed:
change default chunk loader to noop (#3078)

kermandev committed:
fix: Post merge changes

kermandev committed:
refactor: Point methods, docs, etc... (#3007)

* refactor: documentation and improve Point interface methods

* refactor: replace BlockVec instantiation with asBlockVec()

* refactor: improve documentation and clean up PosTest

* refactor: Add similarPoint method and update documentation

* refactor: Add BlockVec and Vec constants and Point documentation

* refactor: Update @Contract annotations in Vec

* refactor: mark deprecated methods with forRemoval and fix asPos method usage in tests

* refactor: update EntityTeleportPacket to use asPos

* refactor: Drop MathUtils usage in coordinate

* feat: Add BlockVec(int, int)

Allows Vec2 components to be easily transitioned to use BlockVec.

So consider a usecase like

new BlockVec(chunkX << 4, chunkY << 4)

* fix: bad merge

* feat: use BlockVec(int, int) for constants

* refactor: Do some final changes

Removes extra line
Removes Easing, in favor of just creating your own with an Operator or using lerp with sampled alpha from EaseFunction.
Make lerpView use float alpha instead of double.
Remove old Pos fixYaw test implementations

kermandev committed:
fix: MetadataDef incorrect index used in AgeableMob

kermandev committed:
feat: Update to 26.1.1

DasBabyPixel committed:
replace chunk synchronized with RW-Lock (#3015)

mudkipdev committed:
Use default entity attributes from vanilla (#2944)

Co-authored-by: Kerman <46640204+kermandev@users.noreply.github.com>

mworzala committed:
chore: rebase

kermandev committed:
fix: Incorrect Generic Bounds & NetworkBufferTemplate (#3043)

kermandev committed:
chore: revert unnecessary rebase change

kermandev committed:
fix: Fluids not working

fluidCount == 0 makes the game skip the section for water, fixes swimming. Though not optimal for the client

ThatGravyBoat committed:
fix: incorrect item particle serialization (#3093)

Bloeckchengrafik committed:
feat: add InventoryButtonClickEvent to handle button interactions in inventories (#3105)

kermandev committed:
refactor!: Delete Query API (#3108)

kermandev committed:
chore: Bump reachability-metadata.json

kermandev committed:
refactor: Post merge changes

Changes Factory to allocator and explicitly makes direct ensure its null terminated.

kermandev committed:
refactor: Delete JDK optimizations.

Can be reintroduced when JDK 27 becomes avaliable in september

kermandev committed:
feat: Add native to NetworkBufferAllocationBenchmark

kermandev committed:
refactor: set readIndex after readOnly

kermandev committed:
chore: Reduce diff

kermandev committed:
refactor: Use byte instead in most cases

kermandev committed:
fix: Regression from reverting JDK 27 changes.

kermandev committed:
chore: Update reachability-meta.json

kermandev committed:
refactor: More changes that revert previous changes, and simplify.

kermandev committed:
refactor: More last changes

kermandev committed:
refactor: Use ArenaStrategy for resizing and trim instead of supplier

Should make the api more extendable and allows for freeing for special ones.

kermandev committed:
chore: Fix docs for arena strategy

kermandev committed:
refactor: Change CachedPacket yet again

Lets just double check lock here. We no longer have the pinning issue since JDK 25, and since we are using virtual threads they can be reallocated to other writers.

Accessing the memory pool is actually quite expensive >200ns for trimmed packets. We should just pay the syncronization cost and let the JVM optimize a single thread likely hitting it in small critical sections.

kermandev committed:
fix: Blocked forever issue with disconnected connections

kermandev committed:
refactor: Bubble up packet errors all the way back

Should make packet misuse actually function now. oops!

kermandev committed:
revert: Functions change

We can revisit this later, not really a requirement at all for this release, and was kind of breaking for no observable improvements yet.

mworzala commented on src/main/java/net/minestom/server/adventure/AdventurePacketConvertor.java:
```diff
@@ -12,39 +11,59 @@
+        ID_NAMED_TEXT_COLOR_MAP = array.toList();
     }
 
+    @ApiStatus.Experimental
```
why experimental?

mworzala commented on src/main/java/net/minestom/server/network/packet/server/play/StopSoundPacket.java:
```diff
@@ -1,35 +1,96 @@
-public record StopSoundPacket(byte flags, @Nullable Sound.Source source,
-                              @Nullable String sound) implements ServerPacket.Play {
-    public static final NetworkBuffer.Type<StopSoundPacket> SERIALIZER = new Type<>() {
+public record StopSoundPacket(Action action) implements ServerPacket.Play {
```
seems very complex for what could just be some factory methods on this class

mworzala commented on src/main/java/net/minestom/server/adventure/MinestomAdventure.java:
```diff
@@ -34,8 +35,10 @@ public final class MinestomAdventure {
      */
-    public static boolean AUTOMATIC_COMPONENT_TRANSLATION = false;
+    @Deprecated(forRemoval = true) // make final to prevent builds that activate it as it would be no op.
+    public static final boolean AUTOMATIC_COMPONENT_TRANSLATION = ServerFlag.AUTOMATIC_COMPONENT_TRANSLATION;
```
its a break anyway so why keep?

mworzala commented on src/main/java/net/minestom/server/entity/metadata/other/PaintingVariant.java:
```diff
@@ -16,8 +16,8 @@ public sealed interface PaintingVariant extends Holder.Direct<PaintingVariant>,
-            NetworkBuffer.COMPONENT.optional(), PaintingVariant::title,
-            NetworkBuffer.COMPONENT.optional(), PaintingVariant::author,
+            NetworkBuffer.OPT_CHAT, PaintingVariant::title,
+            NetworkBuffer.OPT_CHAT, PaintingVariant::author,
```
opt chat is really gross it should be removed and just use optional() or renamed to OPTIONAL_COMPONENT

mworzala commented on src/main/java/net/minestom/server/entity/metadata/villager/VillagerMeta.java:
```diff
@@ -78,15 +78,15 @@ public enum Level {
         }
 
-        public static final NetworkBuffer.Type<Level> NETWORK_TYPE = NetworkBuffer.VAR_INT.transform(Level::fromProtocolId, Level::toProtocolId);
+        public static final NetworkBuffer.Type<Level> NETWORK_TYPE = NetworkBuffer.BYTE.transform(Level::fromProtocolId, Level::toProtocolId);
```
this is wrong: https://mcsrc.dev/1/26.1.2/net/minecraft/world/entity/npc/villager/VillagerData#L39

mworzala commented on src/main/java/net/minestom/server/entity/GameMode.java:
```diff
@@ -37,21 +38,18 @@ public boolean instantBreak() {
-            id -> VALUES[id],
-            gameMode -> (byte) gameMode.ordinal()
-    );
+    public static final NetworkBuffer.Type<@Nullable GameMode> OPT_NETWORK_TYPE = new NetworkBuffer.Type<>() {
```
OPTIONAL_ (ik its not related, but might as well clean it up to be standardized)

mworzala commented on src/main/java/net/minestom/server/network/packet/client/common/ClientResourcePackStatusPacket.java:
```diff
@@ -8,15 +8,15 @@
 ) implements ClientPacket.Configuration, ClientPacket.Play {
     public static final NetworkBuffer.Type<ClientResourcePackStatusPacket> SERIALIZER = NetworkBufferTemplate.template(
             UUID, ClientResourcePackStatusPacket::id,
-            VAR_INT.transform(ClientResourcePackStatusPacket::readStatus, ClientResourcePackStatusPacket::statusId), ClientResourcePackStatusPacket::status,
```
this is not correct. we should not represent the types differently even if it will never encode with multiple bytes

mworzala commented on src/main/java/net/minestom/server/network/packet/client/handshake/ClientHandshakePacket.java:
```diff
@@ -21,7 +21,7 @@ public record ClientHandshakePacket(int protocolVersion, String serverAddress,
             VAR_INT, ClientHandshakePacket::protocolVersion,
             STRING, ClientHandshakePacket::serverAddress,
             UNSIGNED_SHORT, ClientHandshakePacket::serverPort,
-            VAR_INT.transform(Intent::fromId, Intent::id), ClientHandshakePacket::intent,
```
same leave as varint

kermandev commented:
Is being slowly merged in chunks into master instead of one Pull Request.
